src/display/pattern_helper.js

```diff
@@ -322,63 +322,58 @@ var TilingPattern = (function TilingPatternClosure() {
 
       var x0 = bbox[0], y0 = bbox[1], x1 = bbox[2], y1 = bbox[3];
 
-      var topLeft = [x0, y0];
-      // we want the canvas to be as large as the step size
-      var botRight = [x0 + xstep, y0 + ystep];
+      // It is a waste of memory to use the size as specified by xstep or ystep
+      // if the bounding box is smaller.
+      xstep = Math.min(xstep, x1);
+      ystep = Math.min(ystep, y1);
 
-      var width = botRight[0] - topLeft[0];
-      var height = botRight[1] - topLeft[1];
-
-      // Obtain scale from matrix and current transformation matrix.
-      var matrixScale = Util.singularValueDecompose2dScale(this.matrix);
-      var curMatrixScale = Util.singularValueDecompose2dScale(
-        this.baseTransform);
-      var combinedScale = [matrixScale[0] * curMatrixScale[0],
-        matrixScale[1] * curMatrixScale[1]];
-
-      // MAX_PATTERN_SIZE is used to avoid OOM situation.
       // Use width and height values that are as close as possible to the end
       // result when the pattern is used. Too low value makes the pattern look
       // blurry. Too large value makes it look too crispy.
-      width = Math.min(Math.ceil(Math.abs(width * combinedScale[0])),
-        MAX_PATTERN_SIZE);
+      var ctm = this.ctx.mozCurrentTransform;
+      var dimx = this.getSizeAndScale(xstep, this.ctx.canvas.width, ctm[0]);
+      var dimy = this.getSizeAndScale(ystep, this.ctx.canvas.height, ctm[3]);
 
-      height = Math.min(Math.ceil(Math.abs(height * combinedScale[1])),
-        MAX_PATTERN_SIZE);
-
-      var tmpCanvas = CachedCanvases.getCanvas('pattern', width, height, true);
+      var tmpCanvas = CachedCanvases.getCanvas('pattern', dimx.size, dimy.size,
+          true);
       var tmpCtx = tmpCanvas.context;
       var graphics = new CanvasGraphics(tmpCtx, commonObjs, objs);
       graphics.groupLevel = owner.groupLevel;
 
       this.setFillAndStrokeStyleToContext(tmpCtx, paintType, color);
 
-      this.setScale(width, height, xstep, ystep);
-      this.transformToScale(graphics);
-
       // transform coordinates to pattern space
-      var tmpTranslate = [1, 0, 0, 1, -topLeft[0], -topLeft[1]];
-      graphics.transform.apply(graphics, tmpTranslate);
-
+      graphics.transform(dimx.scale, 0, 0, dimy.scale, 0, 0);
```

Comment:
No, I removed it because the transformation actually resulted in a bad rendering: fad83a7deb2d3cb1b6597f11e4e5ecfd7e9a3c94.